File: android/engine/src/test/kotlin/com/zandaulion/omaha/engine/ScoringParityTest.kt
package com.zandaulion.omaha.engine

import kotlinx.coroutines.test.runTest
import java.io.File
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue
import kotlin.time.measureTime

/**
 * Scoring parity — the QuickJS half.
 *
 * The mirror of `test/scoring-parity.test.js` in the repository root. Both run
 * the same `core/scoring.js` against the same fixtures and assert the same
 * output. Any difference between them is *engine* divergence — date parsing,
 * number formatting, `toFixed` rounding — because it is one source file, not
 * two implementations.
 *
 * That is what makes this a measurement rather than an impression, and it is
 * the exit criterion for the QuickJS spike in doc 13 §11.
 *
 * Keep this and the Node test in step. If one grows an allowance the other
 * does not have, the gate stops meaning anything.
 */
class ScoringParityTest {

    private val coreDir = File("../../core").canonicalFile
    private val fixtures = File(coreDir, "__fixtures__")

    private val tickers = listOf("NOK", "AAPL", "JPM")

    private fun fixture(ticker: String, kind: String): String =
        File(fixtures, "$ticker.$kind.json").readText()

    @Test
    fun `core scoring js loads and runs under QuickJS`() = runTest {
        val out = ScoringEngine.create(coreDir).score(fixture("AAPL", "scoring-input"))
        assertTrue(out.isNotBlank(), "engine returned nothing")
        assertTrue(
            out.contains("healthScore"),
            "expected a scored object, got: ${out.take(200)}"
        )
    }

    @Test
    fun `every fixture scores identically to Node`() = runTest {
        // Every ticker is scored before anything is asserted.
        //
        // A loop that asserts inside it stops at the first failure, and the
        // report then says "NOK differs" while saying nothing at all about the
        // other two. That is not a small loss of detail: it was read as "AAPL
        // and JPM pass", recorded in the backlog as such, and used to argue the
        // fault was specific to one ticker's data. All three were failing, for
        // one reason, in every decimal they contained.
        val differed = tickers.filter { ticker ->
            normalise(fixture(ticker, "scoring-output")) !=
                normalise(ScoringEngine.create(coreDir).score(fixture(ticker, "scoring-input")))
        }

        if (differed.isEmpty()) return@runTest

        // The full diff for the first one, since three of them is unreadable,
        // but the roster up front — "all three" and "only NOK" are different
        // diagnoses and the reader needs to know which they are looking at.
        val ticker = differed.first()
        assertEquals(
            normalise(fixture(ticker, "scoring-output")),
            normalise(ScoringEngine.create(coreDir).score(fixture(ticker, "scoring-input"))),
            "Differed from Node: ${differed.joinToString()} (of ${tickers.joinToString()}). " +
                "The diff below is $ticker's.\n\n" +
                "This is engine divergence, not a logic difference — the same " +
                "core/scoring.js produced both. Before reading it as a scoring " +
                "bug, check NumericLocaleTest: a comma decimal separator in the " +
                "host locale corrupts every decimal in the engine and presents " +
                "as exactly this."
        )
    }

    @Test
    fun `a lender still reports Altman Z as inapplicable`() = runTest {
        // The independent claim, restated on this engine. A null that survives
        // Node but not QuickJS would be the exact failure this gate exists for.
        val out = ScoringEngine.create(coreDir).score(fixture("JPM", "scoring-input"))
        assertTrue(
            Regex(""""altmanZ"\s*:\s*null""").containsMatchIn(out),
            "expected altmanZ to be null for a bank; got: ${out.take(300)}"
        )
    }

    @Test
    fun `startup and scoring cost are recorded`() = runTest {
        // Doc 13 §13 lists cold start as unmeasured. This is the measurement.
        val input = fixture("NOK", "scoring-input")

        val coldStart = measureTime { ScoringEngine.create(coreDir).score(input) }

        // Each call builds its own interpreter (see ScoringEngine), so this is
        // the real per-score cost, not a warm-path figure.
        val engine = ScoringEngine.create(coreDir)
        engine.score(input)
        val perScore = measureTime { repeat(10) { engine.score(input) } } / 10

        println("[spike] QuickJS first score (cold JIT)  : $coldStart")
        println("[spike] QuickJS per score (mean of 10)  : $perScore")

        assertTrue(
            coldStart.inWholeMilliseconds < 5_000,
            "cold start took $coldStart, which would be felt on a deep-dive open"
        )
    }

    /**
     * Compare as parsed values rather than as text: key order and whitespace
     * are not part of the contract, and asserting on them would fail for
     * reasons that do not matter.
     */
    private fun normalise(json: String): String = canonical(parseJson(json))
}

File: test/scoring-parity.test.js
/**
 * Scoring parity — the Node reference half.
 *
 * `core/scoring.js` is 1,422 lines and the single riskiest thing to run on a
 * second JavaScript engine. This test states the contract in the narrowest
 * possible terms: given exactly this input, the engine returns exactly this
 * output. No network, no database, no assembly — just the engine.
 *
 * The Android spike runs the identical assertion under QuickJS against the
 * identical files. Any difference between the two is engine divergence, not a
 * logic difference, because it is the same source file. That is what makes the
 * spike a measurement rather than an impression.
 *
 * Keep this test and its Android counterpart in step. If one of them grows an
 * allowance the other does not have, the gate stops meaning anything.
 *
 * Regenerate the fixtures without touching the network:
 *   node scripts/record-fixture.mjs --replay NOK AAPL JPM
 */

import test from 'node:test';
import assert from 'node:assert/strict';
import fs from 'fs';

import { computeComprehensiveHealth } from '../core/scoring.js';
import { fixturePath, normaliseModel } from '../tools/fixture-http.js';

const TICKERS = ['NOK', 'AAPL', 'JPM'];

const read = (ticker, kind) =>
  JSON.parse(fs.readFileSync(fixturePath(ticker, kind), 'utf8'));

for (const ticker of TICKERS) {
  test(`${ticker}: the engine reproduces its recorded output exactly`, () => {
    const input = read(ticker, 'scoring-input');
    const expected = read(ticker, 'scoring-output');

    const actual = normaliseModel(computeComprehensiveHealth(input));

    assert.deepEqual(
      actual,
      expected,
      `${ticker} scoring output drifted. This is the fixture the Android ` +
      'spike compares against — if the change is intended, regenerate with ' +
      '--replay and review the diff.'
    );
  });
}

test('the fixtures cover materially different engine paths', () => {
  const [nok, aapl, jpm] = TICKERS.map((t) => read(t, 'scoring-output'));

  // If all three fixtures exercised the same branches, three of them would be
  // no better than one for the purpose of catching an engine difference.
  assert.equal(jpm.altmanZ ?? null, null, 'JPM is a lender: Altman Z inapplicable');
  assert.ok(typeof aapl.altmanZ === 'number', 'AAPL has a computable Altman Z');

  const scores = [nok, aapl, jpm].map((s) => s.compositeScore ?? s.healthScore ?? null);
  assert.equal(
    new Set(scores).size,
    3,
    `expected three distinct composite scores, saw ${JSON.stringify(scores)}`
  );
});
